File: nim/src/benchmark.nim
import std/[json, times, math, tables, strutils]
import config, helper

type
  Benchmark* = ref object of RootObj

method name*(self: Benchmark): string {.base.} =
  raise newException(ValueError, "Not implemented")

method run*(self: Benchmark, iteration_id: int) {.base.} =
  raise newException(ValueError, "Not implemented")

method checksum*(self: Benchmark): uint32 {.base.} =
  raise newException(ValueError, "Not implemented")

method prepare*(self: Benchmark) {.base.} =
  discard

method config_val*(self: Benchmark, field_name: string): int64 {.base.} =
  config_i64(self.name, field_name)

method iterations*(self: Benchmark): int64 {.base.} =
  self.config_val("iterations")

method expected_checksum*(self: Benchmark): int64 {.base.} =
  self.config_val("checksum")

method warmup_iterations*(self: Benchmark): int64 {.base.} =
  if CONFIG.hasKey(self.name) and CONFIG{self.name}.hasKey("warmup_iterations"):
    return CONFIG{self.name}{"warmup_iterations"}.getInt()
  else:
    let iters = self.iterations
    return max(int64(float(iters) * 0.2), 1'i64)

method warmup*(self: Benchmark) {.base.} =
  let prepare_iters = self.warmup_iterations
  for i in 0..<prepare_iters:
    self.run(i)

proc run_all*(self: Benchmark) =
  let iters = self.iterations
  for i in 0..<iters:
    self.run(i)

proc toLower*(str: string): string =
  result = newString(str.len)
  for i, c in str:
    result[i] = toLowerAscii(c)

type
  BenchmarkFactory* = proc(): Benchmark

var registeredBenchmarks*: Table[string, BenchmarkFactory]

proc registerBenchmark*(name: string, factory: BenchmarkFactory) =
  registeredBenchmarks[name] = factory

proc all*(singleBench = "") =
  var summaryTime = 0.0
  var ok = 0
  var fails = 0

  for name in ORDER:
    if singleBench.len > 0 and name.toLower.find(singleBench.toLower) == -1:
      continue

    let createBenchmark = registeredBenchmarks.getOrDefault(name)
    if createBenchmark == nil:
      echo "Warning: Benchmark '", name, "' defined in config but not found in code"
      continue

    stdout.write(name, ": ")
    stdout.flushFile()

    let bench = createBenchmark()
    reset()
    bench.prepare()

    bench.warmup
    GC_fullCollect()
    reset()

    let start = epochTime()
    bench.run_all
    let duration = epochTime() - start

    if bench.checksum == bench.expected_checksum.uint32:
      stdout.write("OK ")
      inc ok
    else:
      stdout.write("ERR[actual=", $bench.checksum, ", expected=",
          $bench.expected_checksum, "] ")
      inc fails

    echo "in ", formatFloat(duration, ffDecimal, 3), "s"
    summaryTime += duration

    GC_fullCollect()

  if ok + fails > 0:
    echo "Summary: ", formatFloat(summaryTime, ffDecimal, 4), "s, ", ok+fails,
        ", ", ok, ", ", fails

  if fails > 0:
    quit(1)

File: julia/src/Benchmark.jl
abstract type AbstractBenchmark end

function name(b::AbstractBenchmark)::String
    error("Method 'name' not implemented")
end

function run(b::AbstractBenchmark, iteration_id::Int64)
    error("Method 'run' not implemented")
end

function checksum(b::AbstractBenchmark)::UInt32
    error("Method 'checksum' not implemented")
end

function prepare(b::AbstractBenchmark)

end

function warmup(b::AbstractBenchmark)
    warmup_iters = warmup_iterations(b)
    for i = 1:warmup_iters
        run(b, i)
    end
end

function run_all(b::AbstractBenchmark)
    iters = iterations(b)
    for i = 1:iters
        run(b, i)
    end
end

function warmup_iterations(b::AbstractBenchmark)::Int64
    name_str = name(b)
    config = Helper.CONFIG[]

    if haskey(config, name_str) && haskey(config[name_str], "warmup_iterations")
        return config[name_str]["warmup_iterations"]
    else
        iters = iterations(b)
        return max(Int64(round(iters * 0.2)), 1)
    end
end

function iterations(b::AbstractBenchmark)::Int64
    return Helper.config_i64(name(b), "iterations")
end

function expected_checksum(b::AbstractBenchmark)::UInt32
    return UInt32(Helper.config_i64(name(b), "checksum"))
end

function config_val(b::AbstractBenchmark, field_name::String)::Int64
    return Helper.config_i64(name(b), field_name)
end

Benchmark() = Benchmark(0.0)
